File: C++ examples/ChapterFour.cpp
#include <fstream> // Header for file output
#include <vector> // Header for vectors
#include <cmath> // Header for sqrt(), fabs()
#include <ctime> // Header for calculating the runtime
#include <iostream> // Header for output on screen

int main()
{
// Specify simulation parameters
double Theta = -15.0;// Dimensionless applied potential
double Tmax = 1.0;// Dimensionless duration of the potential pulse
double omegaX = 1.05;
double h = 1.e-4;
double omegaT = 1.015;
double deltaT = Tmax / 4000.0;
double Ts = Tmax / 2.0;
clock_t runtime;
runtime = clock();
//Calculate other parameters
// Expanding spatial grid
double maxX = 6.0*sqrt(Tmax);
std::vector<double> X;
X.push_back(0.0);
while(X.back() < maxX)
{
X.push_back(X.back() + h);
h *= omegaX;
}
int n = X.size();// number of spacesteps
//Expanding time grid
std::vector<double> T;
T.push_back(0.0);
while(T.back() < Tmax) // where maxX = 6*sqrt(Tmax)
{
T.push_back(T.back() + deltaT);
if(T.back() > Ts) deltaT *= omegaT;
}
int m = T.size();// number of timesteps
T[m-1] = Tmax;
// Create containers
std::vector<double> alpha(n-1, 0.0);
std::vector<double> beta(n-1, 0.0);
std::vector<double> gamma(n-1, 0.0);
std::vector<double> g_mod(n-2, 0.0);
std::vector<double> delta(n-1, 1.0);
std::vector<double> d_mod(n-1, 0.0);
std::vector<double> C(n, 1.0); // concentration profile
// Calculate Thomas coefficients for T < Ts
double delT = T[1]-T[0];
std::vector<double> delX(n, 0.0);
delX[0] = X[1] - X[0];
for(int i=1; i<n-1; i++)
{
delX[i] = X[i+1] - X[i];
alpha[i] = -(2.0* delT) / ( delX[i-1] * (delX[i-1] + delX[i]) );
gamma[i] = -(2.0* delT) / ( delX[i] * (delX[i-1] + delX[i]) );
beta[i] = 1.0 - alpha[i] - gamma[i];
}
// Modified gamma coefficients for T < Ts
g_mod[0] = 0.0; // boundary condition
for(int i=1; i<n-2; i++) g_mod[i] = gamma[i]
/ (beta[i] - g_mod[i-1] * alpha[i]);
// Open file to output chronoamperogram and concentration profile
std::ofstream Chrono("Chrono_Output.txt");
std::ofstream Profile("Profile_Output.txt");
// BEGIN SIMULATION
for(int k=1; k<m; k++)
{
// Calculate Thomas coefficients for T > Ts
if(T[k] > Ts){
for(int i=1; i<n-1; i++)
{
delT = T[k] - T[k-1];
alpha[i] = -(2.0* delT) / ( delX[i-1] * (delX[i-1] + delX[i]) );
gamma[i] = -(2.0* delT) / ( delX[i] * (delX[i-1] + delX[i]) );
beta[i] = 1.0 - alpha[i] - gamma[i];
}
// Modified gamma coefficients for T > Ts
g_mod[0] = 0.0; // boundary condition
for(int i=1; i<n-2; i++) g_mod[i] = gamma[i]
/ (beta[i] - g_mod[i-1] * alpha[i]);
}
// Forward sweep - create modified deltas
d_mod[0] = 1.0 / (1.0 + exp(-Theta));
for(int i=1; i<n-1; i++) {
d_mod[i] = ( delta[i] - d_mod[i-1]*alpha[i] )
/ ( beta[i] - g_mod[i-1] * alpha[i] );
}
// Back Substitution
C[n-1] = 1.0;
for(int i=n-2; i>=0; i--) {
C[i] = d_mod[i] - g_mod[i] * C[i+1];
delta[i] = C[i];
}
//Output current (and error)
double flux = -(C[1] - C[0])/(X[1] - X[0]);
double reference = - 1.0 / sqrt(M_PI*T[k]);
double QJ = (flux - reference) / reference *100.0;
Chrono << T[k] << "\t" << flux << "\t" << QJ << "\n";
}
// END SIMULATION
runtime = (float)(clock() - runtime)/CLOCKS_PER_SEC*1e3;
std::cout << "Runtime = " << runtime << " ms" << std::endl;
//Output the concentration profile at the end of the pulse
for(int i=0; i<n; i++) Profile << X[i] << "\t" << C[i] << "\n";
}

File: Python examples/ChapterFour.py
import numpy as np

class ChFour:

    def __init__(self, theta = -15, tmax = 1, omegax = 1.05, h = 1E-4, omegat = 1.015):
        
        self.theta = theta
        self.tmax = tmax
        self.omegax = omegax
        self.h = h
        self.omegat = omegat
        
        self.deltat = self.tmax / 4000
        self.ts = self.tmax / 2

        self.maxx = 6 * np.sqrt(self.tmax)
        
        '''Expanding grid'''
        self.x = np.array([0])
        while self.x[-1] < self.maxx:
            self.x = np.append(self.x, self.x[-1] + self.h)
            self.h *= self.omegax

        self.n = int(self.x.size)
        
        '''Expanding time'''
        self.t = np.array([0])
        while self.t[-1] < self.tmax:
            self.t = np.append(self.t, self.t[-1] + self.deltat)
            if self.t[-1] > self.ts:
                self.deltat *= self.omegat

        self.m = int(self.t.size)
        self.t[self.m - 1] = self.tmax
        
        '''Containing arrays'''
        self.alpha = np.zeros((self.n - 1,))
        self.beta = np.zeros((self.n - 1,))
        self.gamma = np.zeros((self.n - 1,))
        self.gmod = np.zeros((self.n - 1,))
        self.delta = np.ones((self.n - 1,))
        self.dmod = np.zeros((self.n - 1,))
        self.C = np.ones((self.n,))

        
        '''Thomas coefficients for T < Ts'''
        self.delt = self.t[1] - self.t[0]
        self.delx = np.zeros((self.n,))
        self.delx[0] = self.x[1] - self.x[0]

        for i in range(1, self.n - 1):
            self.delx[i] = self.x[i + 1] - self.x[i]

            self.alpha[i] = -(2 * self.delt) / (self.delx[i - 1] * (self.delx[i - 1] + self.delx[i]))
            self.gamma[i] = -(2 * self.delt) / (self.delx[i] * (self.delx[i - 1] + self.delx[i]))
            self.beta[i] = 1 - self.alpha[i] - self.gamma[i]
        

        '''Modified gamma coefficients for T < Ts'''   
        self.gmod[0] = 0
        for i in range(1, self.n -1, 1):
            self.gmod[i] = self.gamma[i] / (self.beta[i] - self.gmod[i - 1] * self.alpha[i]) 


        self.time = np.array([])
        self.flux = np.array([])
        self.qj = np.array([])

        for k in range(1, self.m, 1):
            if self.t[k] > self.ts:
                for i in range(1, self.n - 1):
                    self.delt = self.t[k] - self.t[k-1]
                    self.alpha[i] = -(2 * self.delt) / (self.delx[i - 1] * (self.delx[i - 1] + self.delx[i]))
                    self.gamma[i] = -(2 * self.delt) / (self.delx[i] * (self.delx[i - 1] + self.delx[i]))
                    self.beta[i] = 1 - self.alpha[i] - self.gamma[i]
            
                self.gmod[0] = 0
                for i in range(1, self.n - 1):
                    self.gmod[i] = self.gamma[i] / (self.beta[i] - self.gmod[i - 1] * self.alpha[i])
                
            self.dmod[0] = 1 / (1 + np.exp(-self.theta))
            for i in range(1, self.n - 1):
                self.dmod[i] = (self.delta[i] - self.dmod[i - 1] * self.alpha[i]) / (self.beta[i] - self.gmod[i - 1] * self.alpha[i])

            self.C[self.n - 1] = 1
            for i in range(self.n - 2, -1, -1):
                self.C[i] = self.dmod[i] - self.gmod[i] * self.C[i + 1]

                self.delta[i] = self.C[i]

            self.time = np.append(self.time, self.t[k])
            self.flux = np.append(self.flux, -(self.C[1] - self.C[0]) / (self.x[1] - self.x[0]))
            self.ref = -1 / np.sqrt(np.pi * self.t[k])
            self.qj = np.append(self.qj, (self.flux[-1] - self.ref) / (self.ref) * 100)

        self.dist = np.array([])
        self.conc = np.array([])    
        for i in range(0, self.n):
            self.dist = np.append(self.dist, self.x[i])
            self.conc = np.append(self.conc, self.C[i])
        
        self.output = zip(self.time, self.flux, self.qj)
        self.output2 = zip(self.dist, self.conc)
        
    def results(self):
        return(self.output)

if __name__ == '__main__':
    instance = ChFour()
    data = 'C:/Users/SLinf/Documents/data.txt'
    with open(data, 'w') as file:
        for ix, iy, iz in instance.results():
            file.write(str(ix) + ',' + str(iy) + ',' + str(iz) + '\n')